File: 100xDSA/w15-Recursion/CPP/S_Tower_Of_Hanoi.cpp
// ================================
// Author: Pritam
// ================================

#include <iostream>
#include <vector>
#include <algorithm>
#include <string>
//#include <bits/stdc++.h>
using namespace std;

// ---------- Macros & Typedefs ----------
#define ll long long
#define pb push_back
#define all(v) v.begin(), v.end()
#define nl '\n'

// ---------- Solve Function ----------

void f(int n, char src, char aux, char dest) {
    if (n == 0) return;

    f(n-1, src, dest , aux);
    //move the one disk from src to dest
    cout << "Move " << src << " to " << dest << nl;

    f(n-1, aux, src, dest);
}



void solve() {
    int n;
    cin >> n;
    f(n, 'A', 'B', 'C');
}


// ---------- Main ----------
int main() {
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int t = 1;
    // cin >> t;    

    while (t--) {
        solve();
    }

    return 0;
}

File: 100xDSA/w15-Recursion/S_Tower_Of_Hanoi.py
# ================================
# Author: Pritam
# ================================


import sys
input = sys.stdin.readline
# ---------- Helpers ----------
def ceil(a, b):    ##relation b/w floor division and ceil division
    return (a+b-1)//b
def inp():
    return int(input())
def inlt():
    return list(map(int, input().split()))
def ins():
    return input().strip()
def inmap():
    return map(int, input().split())
def print_fast(*args):
    sys.stdout.write(" ".join(map(str, args)) + '\n')
# ---------- Solve Function ----------

def solve():
    n = inp()
    def f(n, src: chr, aux: chr, dest: chr):
        if n == 0:
            return
        ## move all from a to c
        f(n - 1, src, dest, aux)
        print(f"Move {src} to {dest}")
        f(n-1, aux, src, dest)
    f(n, 'A', 'B', 'C')


# ---------- Main ----------
if __name__ == "__main__":
    t = 1
    # t = inp()
    for _ in range(t):
        solve() 